File: source_code/app/services/recommendationService.js
const { Order, Product } = require('../models/schemas');

async function getRecommendations(userId, limit = 6) {
  // Gather categories from past orders
  const orders = await Order.find({ user_id: userId }).lean();
  const boughtIds = new Set();
  const categoryCount = {};

  for (const order of orders) {
    for (const item of order.items) {
      boughtIds.add(item.product_id.toString());
    }
  }

  for (const pid of boughtIds) {
    const product = await Product.findById(pid).lean().catch(() => null);
    if (product) {
      categoryCount[product.category] = (categoryCount[product.category] || 0) + 1;
    }
  }

  const query = {};
  if (Object.keys(categoryCount).length) {
    const topCategories = Object.entries(categoryCount)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 3)
      .map(([cat]) => cat);
    query.category = { $in: topCategories };
  }

  if (boughtIds.size) {
    query._id = { $nin: [...boughtIds] };
  }

  let products = await Product.find(query).sort({ rating: -1 }).limit(limit).lean();

  // Fallback: fill with globally top-rated
  if (products.length < limit) {
    const seenIds = products.map(p => p._id.toString());
    const fallback = await Product.find({ _id: { $nin: seenIds } })
      .sort({ rating: -1 })
      .limit(limit - products.length)
      .lean();
    products = [...products, ...fallback];
  }

  return products.map(p => ({
    id: p._id.toString(),
    name: p.name,
    price: p.price,
    category: p.category,
    image_url: p.image_url,
    rating: p.rating || 0,
  }));
}

module.exports = { getRecommendations };

File: source_code/app/services/recommendation_service.py
from typing import List
from bson import ObjectId

from app.database import get_db


async def get_recommendations(user_id: str, limit: int = 6) -> List[dict]:
    """
    Simple recommendation logic:
    1. Look at user's order history for categories they bought.
    2. Return top-rated products from those categories.
    3. Fall back to globally top-rated products if no history.
    """
    db = get_db()

    # Gather categories from past orders
    orders = await db.orders.find({"user_id": user_id}).to_list(length=50)
    bought_product_ids = set()
    for order in orders:
        for item in order.get("items", []):
            bought_product_ids.add(item["product_id"])

    category_counts: dict = {}
    for pid in bought_product_ids:
        try:
            product = await db.products.find_one({"_id": ObjectId(pid)})
            if product:
                cat = product.get("category", "")
                category_counts[cat] = category_counts.get(cat, 0) + 1
        except Exception:
            continue

    query = {}
    if category_counts:
        top_categories = sorted(category_counts, key=category_counts.get, reverse=True)[:3]
        query["category"] = {"$in": top_categories}
        # Exclude already bought items
        if bought_product_ids:
            try:
                query["_id"] = {"$nin": [ObjectId(pid) for pid in bought_product_ids]}
            except Exception:
                pass

    cursor = db.products.find(query).sort("rating", -1).limit(limit)
    products = await cursor.to_list(length=limit)

    # Fallback: if not enough results, fill with top-rated products globally
    if len(products) < limit:
        seen_ids = {str(p["_id"]) for p in products}
        fallback_cursor = db.products.find({"_id": {"$nin": [p["_id"] for p in products]}}).sort("rating", -1).limit(limit - len(products))
        fallback = await fallback_cursor.to_list(length=limit)
        products.extend(fallback)

    result = []
    for p in products:
        result.append({
            "id": str(p["_id"]),
            "name": p.get("name"),
            "price": p.get("price"),
            "category": p.get("category"),
            "image_url": p.get("image_url"),
            "rating": p.get("rating", 0.0),
        })
    return result
